lusidtools/cocoon/cocoon.py

```diff
@@ -901,11 +904,342 @@ async def _construct_batches(
+            if key in valid_txids:
+                filtered_unmatched_transactions.append(unmatched_transaction)
+
+    return filtered_unmatched_transactions
+
+
+def _unmatched_ids_holdings(
+    api_factory: lusid.utilities.ApiClientFactory,
+    scope: str,
+    sync_batches: list = None,
+):
+    """
+    This method identifies which instruments were not resolved with a holdings upload using load_from_data_frame.
+
+    Parameters
+    ----------
+    api_factory : lusid.utilities.ApiClientFactory api_factory
+        The api factory to use
+    scope : str
+        The scope of the resource to load the data into
+    sync_batches : list
+        A list of the batches used to upload the data into LUSID
+
+    Returns
+    -------
+    responses: list
+        A list to be appended to the ultimate response for load_from_data_frame.
+
+        The returned list will be a unique list of instrument_identifiers.
+
+    """
+    # Extract a list of tuples of portfolio codes and effective at times from sync_batches
+    code_tuples = _extract_unique_portfolio_codes_effective_at_tuples(sync_batches)
+
+    # Create empty list to hold instrument identifiers that have not been resolved
+    unmatched_instruments = []
+
+    # For each holding adjustment in the upload, check whether any contained unresolved instruments and append to list
+    for code_tuple in code_tuples:
+        unmatched_instruments.extend(
+            return_unmatched_instruments_from_holdings(
+                api_factory=api_factory, scope=scope, code_tuple=code_tuple,
+            )
+        )
+
+    # Return a unique list of instrument_identifier dictionaries
+    return list(
+        map(dict, set(tuple(sorted(sub.items())) for sub in unmatched_instruments))
+    )
+
+
+def return_unmatched_instruments_from_holdings(
+    api_factory: lusid.utilities.ApiClientFactory, scope: str, code_tuple: (str, str),
+):
+    """
+    Call the get holdings adjustments api and return a list of instruments that have unresolved identifiers.
+
+    Parameters
+    ----------
+    api_factory : lusid.utilities.ApiClientFactory api_factory
+        The api factory to use
+    scope : str
+        The scope of the resource to load the data into
+    code_tuple : (str, str)
+        A tuple of format (portfolio_code, effective_at) which represents the combination of values
+        used when upserting holdings during load_from_data_frame for file_type 'holding'
+
+    Returns
+    -------
+    A unique list of instrument identifier objects
+
+    """
+    transactions_api = api_factory.build(lusid.api.TransactionPortfoliosApi)
+    unmatched_instruments = []
+
+    response = transactions_api.get_holdings_adjustment(
+        scope=scope, code=code_tuple[0], effective_at=str(DateOrCutLabel(code_tuple[1]))
+    )
+
+    for adjustment in response.adjustments:
+        if adjustment.instrument_uid == "LUID_ZZZZZZZZ":
+            unmatched_instruments.append(adjustment.instrument_identifiers)
```

Comment:
List comprehension should significantly speed this up (list comp vs append)

```
unmatched_instruments = [
    adjustment.instrument_identifiers
    for adjustment in response.adjustments if adjustment.instrument_uid == "LUID_ZZZZZZZZ"
]
```

Reply:
List comprehension works well here 👍 thanks